File: public/packages/dotnet/resilience/Pipeline/CircuitBreakerLayer.cs
// -----------------------------------------------------------------------
// <copyright file="CircuitBreakerLayer.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Resilience.Pipeline;

using DcsvIo.D2.Resilience.CircuitBreaker;

/// <summary>
/// Pipeline layer that fast-fails through a wrapped
/// <see cref="CircuitBreaker{T}"/> when the breaker is open. The per-call
/// key is ignored — circuit breakers operate at the operation level, not
/// the key level.
/// </summary>
/// <typeparam name="TKey">Per-call key type (ignored by this layer).</typeparam>
/// <typeparam name="TValue">The value type produced by the operation.</typeparam>
public sealed class CircuitBreakerLayer<TKey, TValue> : IResilientLayer<TKey, TValue>
    where TKey : notnull
{
    private readonly CircuitBreaker<TValue> r_circuitBreaker;

    /// <summary>
    /// Initializes a <see cref="CircuitBreakerLayer{TKey, TValue}"/>
    /// wrapping the supplied <paramref name="circuitBreaker"/> instance.
    /// </summary>
    /// <param name="circuitBreaker">The CircuitBreaker to wrap.</param>
    public CircuitBreakerLayer(CircuitBreaker<TValue> circuitBreaker)
        => r_circuitBreaker = circuitBreaker;

    /// <inheritdoc/>
    public ValueTask<TValue> WrapAsync(
        TKey key,
        Func<CancellationToken, ValueTask<TValue>> next,
        CancellationToken ct)
        => r_circuitBreaker.ExecuteAsync(next, ct: ct);
}

File: public/packages/typescript/resilience/src/pipeline/circuit-breaker-layer.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import { CircuitBreaker } from "../circuit-breaker/circuit-breaker.js";
import type { CircuitBreakerOptions } from "../circuit-breaker/circuit-breaker-options.js";
import type { IResilientLayer } from "./i-resilient-layer.js";

/**
 * Layer wrapping an inner layer with a per-key circuit breaker. Mirrors .NET
 * `CircuitBreakerLayer<TKey, TValue>` → `CircuitBreaker.ExecuteAsync(next, ct)`.
 *
 * The caller signal threads to the inner op; a caller abort surfaces from the
 * inner op as `AbortError`. As in .NET (`catch { RecordFailure(); throw; }`),
 * the breaker counts an aborted attempt as a failure and re-throws it.
 *
 * The layer calls `breaker.execute(op)` with NO fallback — exactly as the .NET
 * `CircuitBreakerLayer` calls `ExecuteAsync(next, ct)` without one — so an open
 * (or probe-slot-taken) breaker throws `CircuitOpenError` to the pipeline
 * boundary for the caller to map. The breaker's `isFailure` / `onStateChange`
 * seams (when set on the supplied options) thread through unchanged.
 */
export class CircuitBreakerLayer implements IResilientLayer {
  private readonly breakers = new Map<string, CircuitBreaker<unknown>>();

  constructor(
    private readonly inner: IResilientLayer,
    private readonly opts: CircuitBreakerOptions,
  ) {}

  execute<T>(
    key: string,
    op: (signal?: AbortSignal) => Promise<T>,
    signal?: AbortSignal,
  ): Promise<T> {
    let breaker = this.breakers.get(key);
    if (breaker === undefined) {
      breaker = new CircuitBreaker<unknown>(this.opts);
      this.breakers.set(key, breaker);
    }
    return breaker.execute(() =>
      this.inner.execute(key, op, signal),
    ) as Promise<T>;
  }
}
